File: app/src/main/java/me/knighthat/interactivedeck/connection/wireless/WirelessSender.java
package me.knighthat.interactivedeck.connection.wireless;

import android.os.Build;
import android.util.Log;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.io.OutputStream;

public class WirelessSender {

    private static @Nullable OutputStream stream;

    public static void init( @NotNull OutputStream stream ) {
        WirelessSender.stream = stream;
        send(getModel());
        send(getAndroidVer());
    }

    public static void send( @NotNull String msg ) {
        if (stream == null) {
            Log.d("[NET]", "Socket has closed!");
            return;
        }
        try {
            stream.write(msg.getBytes());
            stream.flush();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static @NotNull String getModel() {
        return "device_model: " + Build.MODEL;
    }

    private static @NotNull String getAndroidVer() {
        return "android_version: " + Build.VERSION.RELEASE;
    }
}

File: app/src/main/java/me/knighthat/interactivedeck/connection/wireless/WirelessSender.kt
package me.knighthat.interactivedeck.connection.wireless

import me.knighthat.interactivedeck.connection.request.Request
import me.knighthat.interactivedeck.console.Log
import java.io.IOException
import java.io.OutputStream
import java.util.concurrent.BlockingQueue
import java.util.concurrent.LinkedBlockingQueue

class WirelessSender : Thread() {

    companion object {
        private val QUEUE: BlockingQueue<Request> = LinkedBlockingQueue(20)
        private lateinit var stream: OutputStream

        @JvmStatic
        fun send(request: Request) {
            try {
                QUEUE.put(request)
            } catch (e: InterruptedException) {
                Log.warn("Failed to send request!")
                e.printStackTrace()
            }
        }

        fun start(stream: OutputStream) {
            QUEUE.clear()
            Companion.stream = stream
            WirelessSender().start()
        }
    }


    init {
        name = "NET/O"
    }

    override fun run() {
        while (!interrupted()) {
            try {
                val request = QUEUE.take()
                val serialized = request.toString()

                Log.deb("Sending:")
                Log.deb(serialized)

                stream.write(serialized.toByteArray())
                stream.flush()
            } catch (e: InterruptedException) {
                //TODO Needs proper error handling
                e.printStackTrace()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }
}